runtime/coverage.go

```diff
@@ -20,56 +20,229 @@ package runtime
 func (c *LocationCoverage) AddLineHit(line int) {
 	c.LineHits[line]++
 }
 
-func NewLocationCoverage() *LocationCoverage {
+// Percentage returns a string representation of the covered
+// statements percentage. It is defined as the ratio of covered
+// lines over the total statements for a given location.
+func (c *LocationCoverage) Percentage() string {
+	coveredLines := c.CoveredLines()
+	if coveredLines > c.Statements {
+		// We saturate the percentage at 100%, when the inspector
+		// fails to correctly count all statements for a given
+		// location.
+		coveredLines = c.Statements
+	}
+	percentage := 100 * float64(coveredLines) / float64(c.Statements)
+	return fmt.Sprintf("%0.1f%%", percentage)
+}
+
+// CoveredLines returns the count of covered lines for a given location.
+// This is the number of lines with a hit count > 0.
+func (c *LocationCoverage) CoveredLines() int {
+	coveredLines := 0
+	for _, hits := range c.LineHits { // nolint:maprange
+		if hits > 0 {
+			coveredLines += 1
+		}
+	}
+	return coveredLines
+}
+
+// MissedLines returns an array with the missed lines for a given location.
+// These are all the lines with a hit count == 0. The resulting array is
+// sorted in ascending order.
+func (c *LocationCoverage) MissedLines() []int {
+	missedLines := make([]int, 0)
+	for line, hits := range c.LineHits { // nolint:maprange
+		if hits == 0 {
+			missedLines = append(missedLines, line)
+		}
+	}
+	sort.Ints(missedLines)
+	return missedLines
+}
+
+// NewLocationCoverage creates and returns a *LocationCoverage with the
+// given lineHits map.
+func NewLocationCoverage(lineHits map[int]int) *LocationCoverage {
 	return &LocationCoverage{
-		LineHits: map[int]int{},
+		LineHits:   lineHits,
+		Statements: len(lineHits),
 	}
 }
 
-// CoverageReport is a collection of coverage per location
+// CoverageReport collects coverage information per location.
+// It keeps track of inspected programs per location, and can
+// also exclude locations from coverage collection.
 type CoverageReport struct {
+	// Contains a *LocationCoverage per location.
 	Coverage map[common.Location]*LocationCoverage `json:"-"`
+	// Contains an *ast.Program per location.
+	Programs map[common.Location]*ast.Program `json:"-"`
+	// Contains locations excluded from coverage collection.
+	ExcludedLocations []common.Location `json:"-"`
 }
 
+// ExcludeLocation adds the given location to the array of excluded
+// locations.
+func (r *CoverageReport) ExcludeLocation(location Location) {
+	if r.IsLocationExcluded(location) {
+		return
+	}
+	r.ExcludedLocations = append(r.ExcludedLocations, location)
+}
+
+// IsLocationExcluded checks whether the given location is excluded
+// or not, from coverage collection.
+func (r *CoverageReport) IsLocationExcluded(location Location) bool {
+	for _, excludedLocation := range r.ExcludedLocations {
+		if excludedLocation == location {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Given that `ExcludedLocations` is used only to perform this 'existence' check, maybe use a `map[common.Location]bool` for `ExcludedLocations` instead of an array. Then this lookup would be a lot simpler and faster.
e.g:
```go
func (r *CoverageReport) IsLocationExcluded(location Location) bool {
    return r.ExcludedLocations[location]
} 
```

Reply:
👍 

Even further, maybe just use `map[common.Location]struct{}`, which is effectively the idiomatic way to express a set in Go.

Reply:
Great suggestion, it certainly saved a few lines :100:
Fixed in https://github.com/onflow/cadence/pull/2296/commits/a3d875ede68bcef2b210bfc5b648e94cc0dfaf68